crates/wasi-common/src/sys/windows/path.rs

```diff
@@ -414,13 +414,27 @@ pub(crate) fn filestat_set_times(
 }
 
 pub(crate) fn symlink(old_path: &str, resolved: PathGet) -> Result<()> {
+    use std::fs;
     use std::os::windows::fs::{symlink_dir, symlink_file};
 
     let old_path = concatenate(&resolved.dirfd().as_os_handle(), Path::new(old_path))?;
     let new_path = resolved.concatenate()?;
 
-    // try creating a file symlink
-    let err = match symlink_file(&old_path, &new_path) {
+    // Windows distinguishes between file and directory symlinks.
+    // If the source doesn't exist or is an exotic file type, we fall back
+    // to regular file symlinks.
+    let use_dir_symlink = fs::metadata(&new_path)
+        .as_ref()
+        .map(Metadata::is_dir)
+        .unwrap_or(false);
+
+    let res = if use_dir_symlink {
+        symlink_dir(&old_path, &new_path)
+    } else {
+        symlink_file(&old_path, &new_path)
+    };
```

Comment:
Hmm, I'm somewhat confused here. Why do we need this check? Or put it another way, if we're doing this check already here, is creating a dir symlink on error `ERROR_NOT_A_REPARSE_POINT` still needed? Is the current approach of try and fail and try something else not working?

Some context on why we've originally had `symlink_dir` call after a failing `symlink_file` call. As @sunfishcode explained a long time ago, from atomicity point of view, it seems virtually always better to try and fail and then correct, rather than invoke another syscall (in this case via `fs::metadata`) which would check what the resource (in this case the symlink) is since in between the check and actual call to `symlink_file` etc. the path might change, or whatnot.

Reply:
I was unable to encounter `ERROR_NOT_A_REPARSE_POINT` while using `symlink_file` with the symlink target being a directory. As far as I have tested, the current approach of try and fail doesn't work and a file symlink will always be created.

I agree that try and fail would be preferable, but it appears that `CreateSymbolicLinkA` doesn't check if the symlink target is a file or directory. I don't know if `ERROR_NOT_A_REPARSE_POINT` may be returned by `CreateSymbolicLinkA` at all.

Reply:
Hmm, sounds like a bug then. I thought that `ERROR_NOT_A_REPARSE_POINT` played some role in this, but alas I cannot remember the exact test conditions now. Anyhow, I thought it was covered by our test programs, but from what I understand, you've proved that they were not properly testing this behaviour on Windows? Would the test work fine if we got rid `ERROR_NOT_A_REPARSE_POINT` match altogether? And would they still work without both the match and your upfront check?

Reply:
Oh, and here's another thought. What would happen if we reversed the order of ops, i.e., started with `symlink_dir`, and if failed, tried `symlink_file`. Would that make it possible to apply the "try-catch" approach rather than an upfront check?

Reply:
Trying to create a symlink to a file using `symlink_dir` succeeds too.

Reply:
Interesting, but then the symlinks are obviously invalid if the user would try to act on them in say Windows Explorer or whatnot?

Reply:
Yes, Windows Explorer will show an error, claiming that the directory is invalid.